### PR TITLE
Migrate nym-api HTTP server from rocket to axum

### PR DESCRIPTION
**Pull request by dynco-nym**
# Changes
- start `axum` server alongside `rocket`
- migrate all routes
- write simple API tests that migrated APIs return 200

# TODO
- serve swagger
- migrate logging to tracing
- remove all rocket code

# Notes to reviewers
I left all `rocket` routes alongside new `axum` ones
- `rocket` files are named `routes.rs` or `routes_deprecated.rs`
  + they may show whole file green diffs but it's only rename
- `axum` files are named `handlers.rs`
- merging this into a feature branch: will merge into develop once complete migration is done

Run server
```
cargo run --package nym-api run
```

Simple API tests
```
./nym-node/tests/test_apis.sh
```

Good starting point: compare 
- `start_nym_api_tasks()` to 
- `start_nym_api_tasks_v2()`

### CODE REVIEW

File: common/bin-common/src/logging/mod.rs [hunk 1]
```diff
@@ -41,6 +41,8 @@ pub fn setup_logging() {
         .filter_module("tokio_tungstenite", log::LevelFilter::Warn)
         .filter_module("handlebars", log::LevelFilter::Warn)
         .filter_module("sled", log::LevelFilter::Warn)
+        .filter_module("axum", log::LevelFilter::Info)
+        .filter_module("nym_api", log::LevelFilter::Info)
```
Comment (jstuczyn) [inline common/bin-common/src/logging/mod.rs]: why are we filtering nym-api globally?

File: nym-api/src/main.rs [hunk 2]
```diff
@@ -54,6 +58,7 @@ async fn main() -> Result<(), anyhow::Error> {
     }}
 
     setup_logging();
+    // TODO dz replace with tracing logger once rocket is eliminated from code
```
Comment (jstuczyn) [inline nym-api/src/main.rs:61]: yes please : )

File: nym-api/src/support/http/static_routes.rs
Comment (jstuczyn) [inline nym-api/src/support/http/static_routes.rs:1]: I think this file should live in  the `nym-api-requests` crate as the client will also need to be able to refer to those

File: nym-api/src/support/http/mod.rs [hunk 3]
```diff
@@ -52,33 +63,26 @@ pub(crate) async fn setup_rocket(
         "/network" => network_routes(&openapi_settings),
         "/api-status" => api_status_routes(&openapi_settings),
         "/ecash" => ecash::routes_open_api(&openapi_settings, config.coconut_signer.enabled),
-        "" => nym_node_routes(&openapi_settings),
+        "" => nym_node_routes_deprecated(&openapi_settings),
```
Comment (jstuczyn) [inline nym-api/src/support/http/mod.rs:66]: just for pure curiosity, why did you rename those to deprecated? I mean you're not wrong, because I did just deprecate them on my own branch, but how did you know : )

File: nym-api/src/support/http/mod.rs [hunk 4]
```diff
@@ -154,3 +151,50 @@ fn setup_cors() -> Result<Cors> {
 
     Ok(cors)
 }
+
+#[allow(dead_code)]
+// TODO dz Cannot combine `Access-Control-Allow-Credentials: true` with `Access-Control-Allow-Headers: *`
```
Comment (jstuczyn) [inline nym-api/src/support/http/mod.rs]: imo just remove the allow credentials then because we really don't care about those. we just needed the headers; the reason I had the credentials on my branch is that it was the default behaviour from rocket's cors setup

File: nym-api/src/support/http/mod.rs [hunk 5]
```diff
@@ -154,3 +151,50 @@ fn setup_cors() -> Result<Cors> {
...
+        .route("/", axum::routing::get(|| async {axum::response::Redirect::permanent("/swagger")}))
+        .route("/swagger", axum::routing::get(hello))
+        .merge(circulating_supply_routes())
+        .merge(nym_contract_cache_routes())
+        .merge(node_status_routes(network_monitor))
+        .merge(nym_network_routes())
+        .merge(status::handlers::api_status_routes())
+        .merge(nym_node_routes())
+        .merge(nym_node_routes_unstable())
+        // CORS layer needs to be "outside" of routes
```
Comment (jstuczyn) [inline nym-api/src/support/http/mod.rs:188]: what about the elephant in the room also known as ecash? : )

File: nym-api/src/support/storage/mod.rs [hunk 6]
```diff
@@ -64,6 +64,7 @@ impl NymApiStorage {
         Ok(storage)
     }
 
+    #[deprecated(note = "TODO dz: obsolete because it's used for Rocket")]
     pub(crate) fn stage(storage: NymApiStorage) -> AdHoc {
```
Comment (jstuczyn) [inline nym-api/src/support/storage/mod.rs:68]: you might as well then remove it completely because the only thing this fairing does is to manage the state; so in your temporary rocket builder juts add explicit `rocket.manage(storage)`

File: nym-api/src/support/http/helpers.rs [hunk 7]
```diff
@@ -1,11 +1,72 @@
...
+use tracing::info;
 
 #[derive(Serialize, Deserialize, FromForm, Debug, JsonSchema)]
 pub struct PaginationRequest {
     pub page: Option<u32>,
     pub per_page: Option<u32>,
 }
+
+// TODO dz move this fn to a "common" place so this can be exported ?
+pub async fn logger(
```
Comment (jstuczyn) [inline nym-api/src/support/http/helpers.rs]: I swear I already put this code alongside other common axum helpers in http-common crate on my initial axum-migration branch 🤔

File: nym-api/src/support/config/template.rs [hunk 8]
```diff
@@ -1,6 +1,8 @@
 // Copyright 2021 - Nym Technologies SA <contact@nymtech.net>
 // SPDX-License-Identifier: GPL-3.0-only
 
+// TODO dz any changes to this ?
```
Comment (jstuczyn) [inline nym-api/src/support/config/template.rs:4]: yes. as I had on my branch, you need to put the bind address into the config because rocket pulls it from `Rocket.toml` which obviously won't exist

File: nym-api/src/support/config/mod.rs [hunk 9]
```diff
@@ -234,6 +246,13 @@ pub struct Base {
     #[zeroize(skip)]
     pub local_validator: Url,
 
+    // TODO dz Andrew added this here ?!
```
Comment (jstuczyn) [inline nym-api/src/support/config/mod.rs:249]: I did : ) now we also need it in the template

File: nym-api/src/node_status_api/cache/mod.rs [hunk 10]
```diff
@@ -41,12 +41,13 @@ pub struct NodeStatusCache {
...
 impl NodeStatusCache {
     /// Creates a new cache with no data.
-    fn new() -> NodeStatusCache {
+    pub(crate) fn new() -> NodeStatusCache {
         NodeStatusCache {
             inner: Arc::new(RwLock::new(NodeStatusCacheData::new())),
         }
     }
 
+    #[deprecated(note = "TODO dz: obsolete because it's used for Rocket")]
```
Comment (jstuczyn) [inline nym-api/src/node_status_api/cache/mod.rs:50]: the same comment I had with the storage, just remove it, because it doesn't do much apart from adding it to the managed states

File: nym-api/src/main.rs [hunk 11]
```diff
@@ -107,7 +112,11 @@ async fn start_nym_api_tasks(config: Config) -> anyhow::Result<ShutdownHandles>
     let nym_contract_cache_state = rocket.state::<NymContractCache>().unwrap();
     let node_status_cache_state = rocket.state::<NodeStatusCache>().unwrap();
     let circulating_supply_cache_state = rocket.state::<CirculatingSupplyCache>().unwrap();
-    let maybe_storage = rocket.state::<NymApiStorage>();
+    let storage = if let Some(storage) = rocket.state::<NymApiStorage>() {
```
Comment (jstuczyn) [inline nym-api/src/main.rs:115]: I think that thing can backfire in the future. can we just ensure that the storage always exists when setting up the rocket? though I guess since ultimately we'll remove the rocket, maybe it's fine as the stepping stone

File: nym-api/src/ecash/api_routes/aggregation_axum.rs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+use std::sync::Arc;
+use time::Date;
+
+/// routes with globally aggregated keys, signatures, etc.
+pub(crate) fn aggregation_routes(ecash_state: Arc<EcashState>) -> Router<AxumAppState> {
+    Router::new()
+        .route(
+            "/master-verification-key:epoch_id",
+            axum::routing::get({
+                let ecash_state = Arc::clone(&ecash_state);
```
Comment (jstuczyn) [inline nym-api/src/ecash/api_routes/aggregation_axum.rs:26]: I'm confused, why are we explicitly passing the `EcashState` in an `Arc` rather than using the state extractor like you're already doing with other routes?

Comment (jstuczyn) [reply nym-api/src/ecash/api_routes/aggregation_axum.rs:26]: ah. I think I know why, because under some conditions the routes wouldn't exist... 
🤔 
if it would make your life easier (and I think it would make the code nicer), what do you think of managing `Option<EcashState>`
and then simply at the beginning of each request have something along the lines of 
```rust
let Some(ecash_state) = ecash_state else {
  // return an error response, idk a 400 maybe?
}
```

Comment (jstuczyn) [reply nym-api/src/ecash/api_routes/aggregation_axum.rs:26]: sorry, I'm using rocket wording here ; D
by managing I meant putting it inside your `AxumAppState`

File: nym-api/src/node_status_api/handlers.rs [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+    GatewayCoreStatusResponse, GatewayStatusReportResponse, GatewayUptimeHistoryResponse,
+    GatewayUptimeResponse, InclusionProbabilityResponse, MixNodeBondAnnotated,
+    MixnodeCoreStatusResponse, MixnodeStatusReportResponse, MixnodeStatusResponse,
+    MixnodeUptimeHistoryResponse, RewardEstimationResponse, StakeSaturationResponse,
+    UptimeResponse,
+};
+use nym_mixnet_contract_common::MixId;
+
+pub(crate) fn node_status_routes(network_monitor: bool) -> Router<AxumAppState> {
+    // in the minimal variant we would not have access to endpoints relying on existence
```
Comment (jstuczyn) [inline nym-api/src/node_status_api/handlers.rs:29]: again, if it would make your life and routing easier, perhaps always expose **all** the routes, but simply immediately return an error in the handler if one of the conditions is not fulfilled, such as in this case network monitor not being enabled

Comment (jstuczyn) [reply nym-api/src/node_status_api/handlers.rs:29]: axum routes don't have to behave exactly 100% identical to what we have in rocket

File: common/bin-common/src/logging/mod.rs (hunk 1 above)
Comment (dynco-nym) [reply common/bin-common/src/logging/mod.rs]: for testing, removed

File: nym-api/src/support/http/static_routes.rs
Comment (dynco-nym) [reply nym-api/src/support/http/static_routes.rs:1]: will move it in the next PR (serving swagger docs) where I make some changes to this file

File: nym-api/src/support/http/mod.rs (hunk 4 above)
Comment (dynco-nym) [reply nym-api/src/support/http/mod.rs]: done

File: nym-api/src/support/http/mod.rs (hunk 5 above)
Comment (dynco-nym) [reply nym-api/src/support/http/mod.rs:188]: ecash routes were added in separate commit

File: nym-api/src/support/storage/mod.rs (hunk 6 above)
Comment (dynco-nym) [reply nym-api/src/support/storage/mod.rs:68]: valid point, but I'll leave the function to avoid too much refactoring of rocket code, which will be removed anyway

File: nym-api/src/main.rs (hunk 11 above)
Comment (dynco-nym) [reply nym-api/src/main.rs:115]: Your suggestion over chat was to always init storage: could you elaborate how would it backfire?

File: nym-api/src/node_status_api/handlers.rs (hunk 13 above)
Comment (dynco-nym) [reply nym-api/src/node_status_api/handlers.rs:29]: > always expose all the routes, but simply immediately return an error in the handler if one of the conditions is not fulfilled

Doesn't it make much more sense to not expose routes which shouldn't be there and not return unnecessary 4xx/5xx ?

 It's better from both code and API design perspective

File: nym-api/src/ecash/api_routes/aggregation_axum.rs (hunk 12 above)
Comment (dynco-nym) [reply nym-api/src/ecash/api_routes/aggregation_axum.rs:26]: `Option<State>`  was my initial approach, but... 

Why choose to add an `if` condition for an `Option` you know shouldn't be there, if you can instead encode that info into the type system?

This avoids developer error if a future me or you introduces a change and forgets to initialize an `Option`. 

With my code, it's simply not possible because route either is or isn't there. And if it is, it has access to correct state.

File: nym-api/src/support/http/helpers.rs (hunk 7 above)
Comment (dynco-nym) [reply nym-api/src/support/http/helpers.rs]: yep, moved

File: nym-api/src/main.rs (hunk 11 above)
Comment (octol) [reply nym-api/src/main.rs:115]: I mean will we not remove rocket anyway as part of merging this into develop? Seems like why bother with this then?

**Review by octol — APPROVED** [review]
